packages/commonjs/test/fixtures/function/transpiled-esm-reexported-entry-named/main.js

```diff
@@ -0,0 +1,11 @@
+import * as entry from './proxy';
+
+// Note: There ideally shouldn't be a default generated property.
```

Comment:
Note this particular bug that the change introduces. I'm not sure it's possible to avoid it, but would love suggestions.

Reply:
Hmm, will need to thing about this. In any case your observation

> In order for this to work, we need to check the value __esModule at runtime (unless we prefer to analyze files recursively with Rollup, which appears difficult given Rollup's architecture).

is not correct. Rollup can "look into a module before including it" vie `this.load`, and as a matter of fact, it already does extensively in `resolve-require-sources`. So it should actually be possible to extend this logic to figure out e.g. if we want to create a default export I guess (currently we only know if we are requiring CJS or ES), but this will make things MUCH more complicated.

Reply:
Ah! That's useful info. I'd be willing to explore that if you prefer.

Reply:
Well, have a look, but the logic is slightly involved, feel free to ask questions. I guess the current logic will be good enough as well, but I will need to think a little more about it.

Reply:
After exploring a bit more, I found out that this spurious `default` export is already expected in other tests such as [this one](https://github.com/rollup/plugins/blob/d637611a79a2701c359f5f2f6ffb49978070da38/packages/commonjs/test/fixtures/function/transpiled-esm-entry-named/main.js#L3), so it's not actually new behaviour, and not something we need to worry about in the context of this PR.

I think there is a more general refactor required, which like you suggested would look at the `meta` data collected in required modules and only generate a default export if we expect there to be one. This could even allow us to remove the `getDefaultExportFromCjs` helper entirely (it would always be `export default x["default"]`).

I'm now of the opinion that it would be better done in a separate PR, to be implemented after #1038. It may also be alright to leave it as is :)